New conditions 526

bshyong opened the pull request:
## Description
Adds 'Save' button re: issue https://github.com/department-of-veterans-affairs/vets.gov-team/issues/16468

Original behavior is retained but styling is modified and "Save" button added for more intuitive UX.

Also added `handleSave` action and modified `handleRemove` action

## Testing done
Local testing

## Screenshots
<img width="555" alt="screen shot 2019-02-01 at 12 28 38 am" src="https://user-images.githubusercontent.com/303289/52106429-59a97a00-25b8-11e9-8553-a62adf2243d0.png">


## Acceptance criteria
- [x] Adds save button for visual confirmation when adding new disability

## Definition of done
- [x ] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bshyong committed:
add arrayfield

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
use custom array field

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
custom arrayfield form styling

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
save button

bshyong committed:
remove conditional button display

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
handle save

bshyong committed:
ensure there is a blank item on removal

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

cvalarida commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,427 @@
+        <div className="va-growable">
+          <Element name={`topOfTable_${idSchema.$id}`} />
+          {items.map((item, index) => {
+            // This is largely copied from the default ArrayField
```
This comment comes from the USFS `ArrayField`, which is a bit misleading now, since it kinda sounds like the "default ArrayField" = USFS `ArrayField`, but it really means RJSF's `ArrayField`.

cvalarida commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,427 @@
+                              disabled={!this.props.formData}
+                              onClick={this.handleSave}
+                            >
+                              Save
```
This is the part you added, right?

bshyong replied:
Yep. I added this button. Also added:
- handleSave action
- Modified handleRemove action

cvalarida commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,427 @@
+
+import { scrollToFirstError } from '../shared/utilities/ui';
+import { setArrayRecordTouched } from '../shared/helpers';
+import { errorSchemaIsValid } from '../shared/validation';
```
Rather than copying these files to vets-website, what do you think about importing them from USFS?

cvalarida reviewed (commented):
Can you add a comment somewhere in `ArrayField` that mentions why this is copied over from USFS and what the main changes are?

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
pull in shared funcs from library

bshyong committed:
comments

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong commented:
Updated—the main change here is that a save button and handleSave action are added. The conditional for Edit mode was also changed so that the last item can be in the "confirmed" state—previously the behavior is that the last item is always in the edit state, which is confusing from a UX perspective

cvalarida commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,428 @@
+                            type="button"
+                            onClick={() => this.handleRemove(index)}
+                          >
+                            Remove
```
What do you think of couching this in a `!isOnlyItem` or something so we don't show the remove button if it's the only item on the list (since the remove button won't do anything at that point)?

cvalarida commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,428 @@
+      ),
+    });
+    // Ensure that there is always a blank item in editing mode
+    if (newItems.length === 0) {
```
I'm not sure why this wasn't working, but when I tried to delete the last item, it looks like it didn't update the state properly, but gave me a validation message:

![image](https://user-images.githubusercontent.com/12970166/52369536-ca5ff480-2a05-11e9-9f85-e22ad49b2990.png)

I think if we don't show the remove button on the last item, we won't have to have this condition in here, though.

markgreenburg replied:
I'd vote for not showing a `Remove` button when there's only one item.

bshyong replied:
I agree with getting rid of it; cleaner and makes more sense

markgreenburg commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,428 @@
+        editing: newEditing,
+      });
+
+      this.setState(newState);
```
This looks safe (and is how the other handlers do it) so probably not a real issue but I think [it'd be safer](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous) to pass a function instead since new state is derived from old state. (`newEditing = this.state.editing.map(...`). Maybe something like:

`this.setState((state) => ({ editing: state.editing.map((val, index) => (index + 1 === this.state.editing.length ? false : val)) });`

or more legibly maybe something like:
```
this.setState((state) => {
  const newEditing = state.editing.map(
    (val, index) => (index + 1 === this.state.editing.length ? false : val),
  );
  
  return { editing: newEditing };
}
```

markgreenburg reviewed (commented):
💯  with the exception of the remove button UI. We probably don't want to show it on the last item at all? Might be worth checking with Alex to confirm.

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong committed:
pr feedback; hide remove button on last item

bshyong committed:
conditional flip

bshyong commented:
Updated: here's the new behavior 
http://www.giphy.com/gifs/8OGSmyHbgJNar8xJwu

markgreenburg commented on src/applications/disability-benefits/all-claims/components/ArrayField.jsx:
```diff
@@ -0,0 +1,417 @@
+                          )}
+                        </div>
+                        <div className="small-6 right columns">
+                          {!isOnlyItem && (
```
💯

markgreenburg reviewed (approved):
🚀 awesome!
Only other thing to consider is whether we need any tests for when which buttons show (e.g., test the remove button shows / hides, etc.).

cvalarida reviewed (approved):
Huzzah! @goldenmeanie @katelyncai How does that look to you?

goldenmeanie commented:
LGTM @cvalarida

goldenmeanie commented:
Ahhh, I (maybe) take that back. I don't see an obvious action to take if the edit click was a mistake. Seems like we should probably have a cancel action

cvalarida commented:
:stuck_out_tongue_closed_eyes: I just _had_ to ask! :joy: 

@bshyong What I've done for the cancel button before is save the state when the edit button is clicked, and revert it to that saved state when the cancel button is clicked. This way you don't have to way you don't have to worry about flushing the state to the redux store only when "save" is hit. It gets pretty gnarly when you start mucking with USFS internals.

That said, I _think_ you should be able to do that from the `ArrayField` still, even though the actual state for each item is managed by the `SchemaForm` rendered by `ArrayField`. :crossed_fingers:

bshyong commented:
hmm.. pressing the Save without making any changes is functionally the same as a cancellation but I do see the argument where a user starts editing and then wants to cancel out instead. 

@goldenmeanie @cvalarida how do you see the placement of the cancel button? Should we have the save/cancel group on the left and remove on the right?

cvalarida commented:
I'll leave that decision to @goldenmeanie, but that feels right to me.

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

goldenmeanie commented:
@bshyong I'm happy to chat if you want to talk through any of this :) 

Sorry I didn’t scroll up far enough to see this was your PR!

goldenmeanie commented:
Yeah @bshyong I think Save|Cancel at left, Remove to the right. I think remove could get demoted to a text link.

bshyong committed:
add cancellation behavior

bshyong committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into new-conditions-526

bshyong commented:
cancel button; posting here for record: 

<img width="540" alt="screen shot 2019-02-08 at 11 26 39 am" src="https://user-images.githubusercontent.com/303289/52515660-5891d180-2be4-11e9-8111-3e7c86dd0fd9.png">
